### PR TITLE
Artifact rework

### PR DESCRIPTION
**Pull request by jonasfj**
Integration tests covering larger parts of this will be added later...

### CODE REVIEW

**Comment by jonasfj** [comment]
Fix for the tests was added to another PR... it's just timing robustness... unrelated to changes made here.

File: plugins/artifacts/artifacts.go [hunk 1]
```diff
@@ -1,143 +1,315 @@
...
+		return
 	}
 
-	if len(nonFatalErrs) > 0 {
-		// Only report an exception if the command executed successfully.
-		// The logic behind this it is that no artifact upload failure is
-		// expected for a succeeded run, but failed tasks might be missing
-		// some artifacts.
+	// If resource isn't found, task should fail and we print a message to task log
+	if err == engines.ErrResourceNotFound {
```
Comment (walac) [inline plugins/artifacts/artifacts.go]: Wouldn't this be `err == engines.ErrResourceNotFound && result.Success()`?
If the task has failed, it is expected some artifacts are missing, no reason to fail the plugin.

File: plugins/artifacts/artifacts.go [hunk 2]
```diff
@@ -1,143 +1,315 @@
...
+	if err == runtime.ErrNonFatalInternalError {
+		tp.nonFatalErr.Set(true) // Propagate the non-fatal error
+		tp.monitor.Warn("Received ErrNonFatalInternalError from ResultSet.ExtractFile()")
+		return
+	}
+
+	// If fatal internal error, we want to propagate (what else can we do)
+	if err == runtime.ErrFatalInternalError {
+		tp.fatalErr.Set(true) // Propagate the fatal error
+		tp.monitor.Error("Received ErrFatalInternalError from ResultSet.ExtractFile()")
```
Comment (walac) [inline plugins/artifacts/artifacts.go]: If `err` contains useful information, we should log it here

File: plugins/artifacts/artifacts.go [hunk 3]
```diff
@@ -1,143 +1,315 @@
...
+		))
+		tp.mErrors.Unlock()
+		return
+	}
 
-		return false, nil
+	// If non-fatal internal error, we want to propagate (what else can we do)
+	if err == runtime.ErrNonFatalInternalError {
+		tp.nonFatalErr.Set(true) // Propagate the non-fatal error
+		tp.monitor.Warn("Received ErrNonFatalInternalError from ResultSet.ExtractFile()")
```
Comment (walac) [inline plugins/artifacts/artifacts.go]: If `err` contains useful information, we should log it here

File: plugins/artifacts/artifacts.go [hunk 4]
```diff
@@ -1,143 +1,315 @@
...
+		tp.mErrors.Lock()
+		tp.errors = append(tp.errors, e)
+		tp.mErrors.Unlock()
+		return
 	}
 
-	if reason != "" {
-		tp.context.Log("Artifact upload error handled. Continuing...", name, err.Error())
+	// If resource isn't found, task should fail and we print a message to task log
+	if err == engines.ErrResourceNotFound {
```
Comment (walac) [inline plugins/artifacts/artifacts.go]: See my comment in `processFile`

File: plugins/artifacts/artifacts.go [hunk 5]
```diff
@@ -1,143 +1,315 @@
...
+		tp.mErrors.Unlock()
+		return
+	}
 
-func init() {
-	plugins.Register("artifacts", pluginProvider{})
+	// If non-fatal internal error, we want to propagate (what else can we do)
+	if err == runtime.ErrNonFatalInternalError {
+		tp.nonFatalErr.Set(true) // Propagate the non-fatal error
+		tp.monitor.Warn("Received ErrNonFatalInternalError from ResultSet.ExtractFolder()")
```
Comment (walac) [inline plugins/artifacts/artifacts.go]: Same comment about logging `err`

File: plugins/artifacts/artifacts.go [hunk 6]
```diff
@@ -1,143 +1,315 @@
...
+		return
+	}
+
+	// If fatal internal error, we want to propagate (what else can we do)
+	if err == runtime.ErrFatalInternalError {
+		tp.fatalErr.Set(true) // Propagate the fatal error
+		tp.monitor.Error("Received ErrFatalInternalError from ResultSet.ExtractFolder()")
```
Comment (walac) [inline plugins/artifacts/artifacts.go]: Same comment about logging `err`

File: plugins/artifacts/doc.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+// Package artifacts is provides a taskcluster-worker plugin that uploads
```
Comment (walac) [inline plugins/artifacts/doc.go]: s/is //

File: worker/workertest/artifactassertions.go [hunk 8]
```diff
@@ -20,6 +21,21 @@ func GrepArtifact(substring string) func(t *testing.T, a Artifact) {
 	}
 }
 
+// LogArtifact creates an assetions that logs the artifact, to test log.
```
Comment (walac) [inline worker/workertest/artifactassertions.go]: s/assetions/assertion/

File: plugins/artifacts/artifacts.go (hunk 2 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: It doesn't... It just is a specific value...

File: plugins/artifacts/artifacts.go (hunk 1 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: We should set failed regardless... of `result.Success()` (just for good measure)...
I would argue the same for the error artifact...

Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: regardless, we definitely need to handle the error... so it's not reported as unhandled error.

File: plugins/artifacts/artifacts.go (hunk 3 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: it doesn't... `err == runtime.ErrNonFatalInternalError`, ie. it's a specific value.

File: plugins/artifacts/artifacts.go (hunk 4 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: Okay, I'm moving the `CreateErrorArtifact` under the `if result.Success() {`...
but I do want to handle the error...

File: plugins/artifacts/artifacts.go (hunk 6 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: contains no info

File: plugins/artifacts/artifacts.go (hunk 5 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: contains no info

File: plugins/artifacts/artifacts.go (hunk 4 above)
Comment (jonasfj) [inline plugins/artifacts/artifacts.go]: okay, changed my mind again...

**Pull request merged**